From nicolas.ferre@atmel.com  Thu Apr 29 16:28:30 2010
From: Nicolas Ferre <nicolas.ferre@atmel.com>
Date: Wed, 28 Apr 2010 13:45:40 +0200
Subject: USB: ohci-at91: fix power management hanging
To: linux-usb@vger.kernel.org, avictor.za@gmail.com, david-b@pacbell.net
Cc: linux-kernel@vger.kernel.org, linux-arm-kernel@lists.infradead.org, Patrice Vilchez <patrice.vilchez@atmel.com>
Message-ID: <1272455140-32686-1-git-send-email-nicolas.ferre@atmel.com>


From: Patrice Vilchez <patrice.vilchez@atmel.com>

A hanging has been detected in ohci-at91 while going in suspend to ram. This is
due to asynchronous operations between ohci reset and ohci clocks shutdown.
This patch adds the reading of the control register between the reset of the
ohci and clocks stop. This "flush the writes" idea was taken from ohci-hcd.c
file (ohci_shutdown() function).

Signed-off-by: Patrice Vilchez <patrice.vilchez@atmel.com>
Signed-off-by: Nicolas Ferre <nicolas.ferre@atmel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/host/ohci-at91.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/usb/host/ohci-at91.c
+++ b/drivers/usb/host/ohci-at91.c
@@ -331,6 +331,8 @@ ohci_hcd_at91_drv_suspend(struct platfor
 	 */
 	if (at91_suspend_entering_slow_clock()) {
 		ohci_usb_reset (ohci);
+		/* flush the writes */
+		(void) ohci_readl (ohci, &ohci->regs->control);
 		at91_stop_clock();
 	}
 
